core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -195,12 +195,14 @@ private void initialize()
             if (!state.canScheduleMoreTasks()) {
                 taskLifecycleListener.noMoreTasks(stage.getFragment().getId());
 
-                // update output buffers
-                for (PlanFragmentId sourceFragment : exchangeSources.keySet()) {
-                    OutputBufferManager outputBufferManager = outputBufferManagers.get(sourceFragment);
-                    outputBufferManager.noMoreBuffers();
-                    for (RemoteTask sourceTask : sourceTasks.get(stage.getFragment().getId())) {
-                        sourceTask.setOutputBuffers(outputBufferManager.getOutputBuffers());
+                synchronized (this) {
```

Comment:
Was this a new code or copied from somewhere?

Reply:
new. It was vastly restructured.